## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [269]:
import pandas as pd
from decimal import Decimal, ROUND_HALF_UP

df = pd.read_csv('./Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [270]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [271]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [272]:
# Use this space to do any additional research
# and familiarize yourself with the data.
top_category = df['category'].value_counts().idxmax()
top_category



'consumables'

In [273]:
# What three item categories had the most entries?
# Use the value_counts method to find the count of each category
# Sort the results in descending order
# Use the head function to limit the output to the top 3 categories
top_three = df['category'].value_counts().head(3)
top_three



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [274]:
# For the category with the most entries,
# which subcategory had the most entries?
# Create a new DataFrame filtered by the top category
# Use the value_counts method to find the count of each subcategory
# Sort the results in descending order
# Use the head function to limit the output to the top subcategory
top_subcategory = df.loc[df['category'] == top_category]
top_subcategory = top_subcategory['subcategory'].value_counts().head(1)
top_subcategory



subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [275]:
# Which five clients had the most entries in the data?
# Use the value_counts method to find the count of each client
# Sort the results in descending order
# Use the head function to limit the output to the top 5 clients
top_five = df['client_id'].value_counts().head(5)
top_five



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [276]:
# Store the client ids of those top 5 clients in a list.
# Use the tolist method to convert the index to a list
top_five_list = top_five.index.tolist()
top_five_list



[33615, 66037, 46820, 38378, 24741]

In [277]:
# How many total units (the qty column) did the
# client with the most entries order order?
# Create a new DataFrame filtered by the top client
# Use the sum method to get the sum of the qty column
top_client = df.loc[df['client_id'] == top_five_list[0]]
top_client = top_client['qty'].sum()
top_client



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [278]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df['line_subtotal'] = df['line_subtotal'].apply(Decimal).apply(lambda x: x.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number line_subtotal  
0     1096.80     762.71         7.50  105            1     115164.00  
1       24.95      15.09         1.49   21            0        523.95  
2       13.52       7.86         1.68   39            6        527.28  
3       36.42      24.85         1.23   29            3       1056.18  
4      195.10     108.17        46.43   20            1       3902.00

In [279]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
import numpy as np
df['shipping_price'] = np.where((df['unit_weight']*df['qty']) > 50, (df['unit_weight']*df['qty'])  * 7, (df['unit_weight']*df['qty']) * 10)
df['shipping_price'] = df['shipping_price'].apply(Decimal).apply(lambda x: x.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number line_subtotal  \
0     1096.80     762.71         7.50  105            1     115164.00   
1       24.95      15.09         1.49   21            0        523.95   
2       13.52       7.86         1.68   39            6        527.28   
3       36.42      24.85         1.23   29            3       1056.18   
4      195.10     108.17        46.43   20            1       3902.00   

  shipping_price  
0        5512.50  
1         312.90  
2         458.64  
3         356.70  
4        6500.20

In [280]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
# Let's start by adding a sales tax column storing sales tax as a float
df['sales_tax'] = Decimal('0.0925')
# We will now add the total price column
df['total_price'] = ((df['line_subtotal'] + df['shipping_price']) * (1 + df['sales_tax']))
df['total_price'] = ((df['line_subtotal'] + df['shipping_price']) * (1 + df['sales_tax'])).apply(lambda x: x.quantize(Decimal('.01'), rounding=ROUND_HALF_UP))
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  shipping_price sales_tax  \
0         7.50  105            1      115164.00         5512.50    0.0925   
1         1.49   21            0         523.95          312.90    0.0925   
2         1.68   39            6         527.28          458.64    0.0925   
3         1.23   29            3        1056.18          356.70    0.0925   
4        46.43   20            1        3902.00         6500.20    0.0925   

  total_price  
0   131839.08  
1      914.26  
2     1077.12  
3     1543.57  
4    11364.40  

[5 rows x 22 columns]

In [281]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['unit_cost'] = df['unit_cost'].apply(Decimal).apply(lambda x: x.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df['line_cost'] = df['line_cost'].apply(Decimal).apply(lambda x: x.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

  line_number  line_subtotal  shipping_price  sales_tax total_price line_cost  
0           1      115164.00         5512.50     0.0925   131839.08  85597.05  
1           0         523.95          312.90     0.0925      914.26    629.79  
2           6         527.28          458.64     0.0925     1077.12    765.18  
3           3        1056.18          356.70     0.0925     1543.57   1077.35  
4           1        3902.00         6500.20     0.0925    11364.40   8663.60  

[5 rows x 23 columns]

In [282]:
# Create a column for the profit of
# each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df.head()





first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

  line_subtotal  shipping_price  sales_tax  total_price line_cost line_profit  
0     115164.00         5512.50     0.0925    131839.08  85597.05    46242.03  
1        523.95          312.90     0.0925       914.26    629.79      284.47  
2        527.28          458.64     0.0925      1077.12    765.18      311.94  
3       1056.18          356.70     0.0925      1543.57   1077.35      466.22  
4       3902.00         6500.20     0.0925     11364.40   8663.60     2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [283]:
# Check your work using the totals above
# to make sure they match up
# group the data by order ID and sum the total price for each order
order_totals = df.groupby('order_id')['total_price'].sum()

# compare the results with the given receipts
expected_totals = {2742071: 152811.89, 2173913: 162388.71, 6128929: 923441.25}
for order_id, expected_total in expected_totals.items():
    calculated_total = df[df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id} - Expected: ${expected_total}, Calculated: ${calculated_total}")




Order ID 2742071 - Expected: $152811.89, Calculated: $152811.89
Order ID 2173913 - Expected: $162388.71, Calculated: $162388.71
Order ID 6128929 - Expected: $923441.25, Calculated: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [284]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
# Create a new DataFrame filtered by the top 5 clients
# Use the groupby method to group the data by client id
# Use the sum method to get the sum of the total price column
top_five_totals = df.loc[df['client_id'].isin(top_five_list)].groupby('client_id')['total_price'].sum()
top_five_totals



client_id
24741    82268892.04
33615     8377308.55
38378    12906550.87
46820     9743794.36
66037    10259514.80
Name: total_price, dtype: object

In [285]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
# Hint: Use the agg method
summary_df = df.loc[df['client_id'].isin(top_five_list)].groupby('client_id').agg({'qty': 'sum', 'shipping_price': 'sum', 'total_price': 'sum', 'line_profit': 'sum'})
summary_df = summary_df.rename(columns={'qty': 'total_units_purchased', 'shipping_price': 'total_shipping_price', 'total_price': 'total_revenue', 'line_profit': 'total_profit'})
summary_df = summary_df.sort_values('total_profit', ascending=False)
summary_df




total_units_purchased total_shipping_price total_revenue  \
client_id                                                             
24741                     239862           5126448.37   82268892.04   
38378                      73667           3429455.40   12906550.87   
66037                      43018           1395151.85   10259514.80   
46820                      75768           1601448.84    9743794.36   
33615                      64313           1828984.89    8377308.55   

          total_profit  
client_id               
24741      36579992.33  
38378       3271829.89  
66037       3255031.82  
46820       2736602.72  
33615       2201994.64

In [286]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
# Hint: Use the apply method
summary_df['total_shipping_price'] = summary_df['total_shipping_price'].map("${:.2f}".format)
summary_df['total_revenue'] = summary_df['total_revenue'].map("${:.2f}".format)
summary_df['total_profit'] = summary_df['total_profit'].map("${:.2f}".format)
summary_df['total_units_purchased'] = summary_df['total_units_purchased'].map("{:,}".format)
summary_df



total_units_purchased total_shipping_price total_revenue  \
client_id                                                            
24741                   239,862          $5126448.37  $82268892.04   
38378                    73,667          $3429455.40  $12906550.87   
66037                    43,018          $1395151.85  $10259514.80   
46820                    75,768          $1601448.84   $9743794.36   
33615                    64,313          $1828984.89   $8377308.55   

           total_profit  
client_id                
24741      $36579992.33  
38378       $3271829.89  
66037       $3255031.82  
46820       $2736602.72  
33615       $2201994.64

In [287]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df = summary_df.sort_values('total_profit', ascending=False)
summary_df



total_units_purchased total_shipping_price total_revenue  \
client_id                                                            
24741                   239,862          $5126448.37  $82268892.04   
38378                    73,667          $3429455.40  $12906550.87   
66037                    43,018          $1395151.85  $10259514.80   
46820                    75,768          $1601448.84   $9743794.36   
33615                    64,313          $1828984.89   $8377308.55   

           total_profit  
client_id                
24741      $36579992.33  
38378       $3271829.89  
66037       $3255031.82  
46820       $2736602.72  
33615       $2201994.64

### Summary of E-Commerce Data Analysis

In this comprehensive analysis of the e-commerce dataset, we've gained valuable insights into customer behavior and business performance:

1. **Key Product Categories and Clients**: The analysis revealed that the 'consumables' category, particularly 'bathroom supplies,' was highly favored among the customers. This insight can guide inventory stocking and marketing strategies. Additionally, we identified the top five clients by the number of entries, with client IDs 24741, 33615, 38378, 46820, and 66037, contributing significantly to the business.

2. **Financial Insights**: Our deep dive into the financial aspects of the transactions highlighted the substantial spending by these top clients. The total spendings were as follows: client 24741 spent approximately $82.27 million, client 33615 spent about $8.38 million, client 38378 spent around $12.91 million, client 46820 spent nearly $9.74 million, and client 66037 spent about $10.26 million. These figures are instrumental in understanding the revenue distribution and prioritizing client relationships.

3. **Strategic Implications**: The findings from this analysis are crucial for shaping our strategic decisions. Emphasizing product categories with higher demand and nurturing relationships with top clients can lead to increased sales and customer loyalty. Furthermore, this analysis serves as a benchmark for future business reviews and data-driven decision-making.

Overall, this project not only provided a clear picture of the current business scenario but also laid down a pathway for future growth strategies based on real data insights.
